src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,215 @@
+  const std::string& name() const { return name_; }
+
+ private:
+  std::string name_;
+};
+
+// Used to represent whether a predicate is negated or not
+enum Sign { kNegated, kPositive };
+
+// Predicate corresponds to a predicate in of the form 
+// <pred_name>(arg_1, ..., arg_n), and it may or may not be negated
+class Predicate {
+ public:
+  Predicate(std::string name, std::vector<std::string> args, Sign sign)
+      : name_(std::move(name)),
+        args_(std::move(args)),
+        // TODO move surpresses an unused private field error about sign_
+        // for now, get rid of this eventually
+        sign_(std::move(sign)) {}
+
+  const std::string& name() const { return name_; }
+  const std::vector<std::string>& args() const { return args_; }
+  Sign sign() { return sign_; }
+
+ private:
+  std::string name_;
+  std::vector<std::string> args_;
+  Sign sign_;
+};
+
+// Attribute corresponds to an attribute of a principal with the form
+// <Principal> <pred_name>(arg_1, ..., arg_n)
+class Attribute {
+ public:
+  explicit Attribute(Principal principal, Predicate predicate)
+    : principal_(principal), predicate_(predicate) {}
+  const Principal& principal() const { return principal_; }
+  const Predicate& predicate() const { return predicate_; }
+ private:
+  Principal principal_;
+  Predicate predicate_;
+};
+
+// CanActAs corresponds to a canActAs expression of the form
+// <principal_1> canActAs <principal_2>
+class CanActAs {
+ public:
+  explicit CanActAs(Principal left_principal, Principal right_principal)
+    : left_principal_(std::move(left_principal)),
+      right_principal_(std::move(right_principal)) {}
+  const Principal& left_principal() const { return left_principal_; }
+  const Principal& right_principal() const { return right_principal_; }
+
+ private:
+   Principal left_principal_;
+   Principal right_principal_;
+};
+
+// BaseFact corresponds to a base-case fact that cannot recursively include other 
+// facts through "canSay". "Fact" includes the recursive case. It is necessary 
+// that BaseFact is a separate class from Fact because the RHS of conditional 
+// assertions can only contain BaseFacts.
+// Base facts have three forms:
+//  - A predicate
+//  - An attribute
+//  - A canActAs expression
+// The particular choice of form is represented with a variant which is 
+// encapsulated.
+class BaseFact {
+ public:
+  explicit BaseFact(Predicate predicate) : value_(std::move(predicate)) {};
+  explicit BaseFact(Attribute attribute) : value_(std::move(attribute)) {};
+  explicit BaseFact(CanActAs canActAs) : value_(std::move(canActAs)) {};
+
+ private:
+  std::variant<Predicate, Attribute, CanActAs> value_;
+};
+
+// CanSay and Fact are forward-declared because they are mutually recursive.
+class CanSay;
+class Fact;
+
+// CanSay corresponds to an expression of the form <principal> canSay Fact
+class CanSay {
+  public:
+    explicit CanSay(Principal principal, std::unique_ptr<Fact>& fact)
```

Comment:
This should be a `std::unique_ptr<Fact>` rather than a reference to the `std::unique_ptr<Fact>`. The `std::move` call will force the old name to be erased when being placed in `fact_`, thus allowing only one live unique pointer to exist at a time.